### PR TITLE
No more gossip flood

### PR DESCRIPTION
**Pull request by rustyrussell**
This gets us a long way towards #1879; it can do with further refinement, but it's the minimal fix (important for the release , since upgrading will discard the store!).

~It's built on top of #2685 and #2644 so review after 724b0cf~ (merged)

### CODE REVIEW

File: common/read_peer_msg.c [hunk 1]
```diff
@@ -117,23 +117,35 @@ void handle_gossip_msg(struct per_peer_state *pps, const u8 *msg TAKES)
...
-		peer_failed_connection_lost();
+		sync_crypto_write(pps, gossip);
 	}
 
 out:
 	if (taken(msg))
 		tal_free(msg);
 }
 
+/* takes iff returns true */
```
[the remark this answers is not in the document]
Comment (niftynei) [reply common/read_peer_msg.c]: `iff` is shorthand for 'if and only if'

[the remark this answers is not in the document]
Comment (darosior) [reply common/read_peer_msg.c]: Thanks didnt know that :sweat_smile:

**Comment by rustyrussell** [comment]
Rebased, no changes.

File: gossipd/gossipd.c [hunk 2]
```diff
@@ -1669,6 +1714,35 @@ static struct io_plan *peer_msg_in(struct io_conn *conn,
...
+	list_for_each(&daemon->peers, peer, list)
+		gossip_levels[peer->gossip_level]++;
+
+	/* If we're missing gossip, try to fill GOSSIP_HIGH */
+	if (daemon->gossip_missing != NULL)
+		glevel = GOSSIP_HIGH;
+	else
+		glevel = GOSSIP_MEDIUM;
+
+	while (gossip_levels[glevel] > gossip_level_targets[glevel])
```
Comment (cdecker) [inline gossipd/gossipd.c]: Shouldn't this be `>=`?

Comment (rustyrussell) [reply gossipd/gossipd.c]: Yes, good point.  We also don't provide any insight into what gossip level is for each peer, which would be nice (but means a round-trip to gossipd for listpeers...).

Fixed

**Comment by rustyrussell** [comment]
Fix from @cdecker and trivial rebase to prevent CHANGELOG reject.

**Pull request merged**